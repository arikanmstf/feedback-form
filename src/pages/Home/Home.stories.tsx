import { ComponentStory, ComponentMeta } from '@storybook/react';

import handlers from '../../mocks/handlers';

import Home from './Home';

export default {
  title: 'Home',
  component: Home,
} as ComponentMeta<typeof Home>;

const Template: ComponentStory<typeof Home> = () => <Home />;

export const Default = Template.bind({});
Default.parameters = {
  msw: {
    handlers,
  },
};
