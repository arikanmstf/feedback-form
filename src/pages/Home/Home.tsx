import SendIcon from '@mui/icons-material/ChevronRight';
import { Button, TextField } from '@mui/material';
import Head from 'next/head';
import { useRouter } from 'next/router';
import { useState } from 'react';
import { useForm, Controller } from 'react-hook-form';

import {
  Page,
  ColumnContainer,
  RowContainer,
  RightAlignedContainer,
  StarInput,
} from '../../components';
import HeadLine from '../../components/HeadLine';
import { createComment } from '../../services/comment';
import { validateEmail } from '../../utils/validation';

const Home = () => {
  // hooks
  const [loading, setLoading] = useState(false);
  const {
    control,
    handleSubmit,
    formState: { errors: formErrors },
  } = useForm({
    defaultValues: {
      name: '',
      email: '',
      content: '',
      stars: 0,
    },
  });
  const { push } = useRouter();

  // callbacks
  const onPost = async (data) => {
    setLoading(true);
    try {
      await createComment(data);
      await push('/comments');
    } catch (e) {
      // tools like rollbar automatically reports console.error
      // eslint-disable-next-line no-console
      console.error(e);
    } finally {
      setLoading(false);
    }
  };
  const title = 'Share your feedback with us!';

  return (
    <>
      <Head>
        <title>{title}</title>
      </Head>
      <Page>
        <HeadLine label={title} />
        <ColumnContainer>
          <RowContainer>
            <Controller
              control={control}
              rules={{ required: true }}
              render={({ field: { onChange, value } }) => (
                <TextField
                  error={!!formErrors.name}
                  onChange={onChange}
                  value={value}
                  required
                  id="name"
                  label="Name"
                  helperText={formErrors.name ? 'Please enter your name' : undefined}
                />
              )}
              name="name"
            />

            <Controller
              control={control}
              rules={{ required: true, validate: validateEmail }}
              render={({ field: { onChange, value } }) => (
                <TextField
                  error={!!formErrors.email}
                  onChange={onChange}
                  value={value}
                  required
                  id="email"
                  label="Email"
                  helperText={formErrors.email ? 'Please enter a valid email address' : undefined}
                />
              )}
              name="email"
            />

            <Controller
              control={control}
              rules={{ required: true, validate: (nr) => +nr > 0 }}
              render={({ field: { onChange, value } }) => (
                <StarInput
                  onChange={onChange}
                  value={value}
                  error={!!formErrors.stars}
                  helperText={formErrors.stars ? 'Please select some stars' : undefined}
                />
              )}
              name="stars"
            />
          </RowContainer>

          <RowContainer>
            <div>
              <Controller
                control={control}
                rules={{ required: true }}
                render={({ field: { onChange, value } }) => (
                  <TextField
                    onChange={onChange}
                    value={value}
                    error={!!formErrors.content}
                    style={{ alignSelf: 'stretch', width: '100%' }}
                    required
                    multiline
                    id="content"
                    rows={6}
                    label="Your comment"
                    helperText={formErrors.content ? 'Please enter your comment' : undefined}
                  />
                )}
                name="content"
              />
            </div>
          </RowContainer>
        </ColumnContainer>
        <RightAlignedContainer>
          <Button
            aria-label={loading ? 'Posting' : 'Post'}
            onClick={handleSubmit(onPost)}
            disabled={loading}
            endIcon={<SendIcon />}
            style={{ minWidth: 128 }}
            variant="contained"
          >
            {loading ? 'Posting' : 'Post'}
          </Button>
        </RightAlignedContainer>
      </Page>
    </>
  );
};

export default Home;
