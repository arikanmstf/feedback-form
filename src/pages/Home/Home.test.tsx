import renderWithProviders from '../../test/utils/render';

import Home from './Home';

describe('Home', () => {
  it('should render correctly', () => {
    const { container } = renderWithProviders(<Home />);
    expect(container).toMatchSnapshot();
  });
});
