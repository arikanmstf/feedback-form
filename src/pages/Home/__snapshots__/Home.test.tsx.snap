// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<div>
  <div
    class="sc-eDvSVe bYLQkd"
  >
    <div
      class="sc-hLBbgP hIciRX"
    >
      <h4
        aria-label="Share your feedback with us!"
        class="MuiTypography-root MuiTypography-h4 css-4x1bza-MuiTypography-root"
      >
        Share your feedback with us!
      </h4>
    </div>
    <div
      class="sc-dkrFOg kSKhVS"
    >
      <div
        class="sc-gKPRtg lPbDl"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1vk6yy5-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="name"
            id="name-label"
          >
            Name
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-fbz57p-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="name"
              required=""
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Name
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1vk6yy5-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="email"
            id="email-label"
          >
            Email
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-fbz57p-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="email"
              required=""
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Email
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <section>
          <button
            aria-label="Give 1 stars"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="StarsIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Give 2 stars"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="StarsIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Give 3 stars"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="StarsIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Give 4 stars"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="StarsIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Give 5 stars"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="StarsIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zm4.24 16L12 15.45 7.77 18l1.12-4.81-3.73-3.23 4.92-.42L12 5l1.92 4.53 4.92.42-3.73 3.23L16.23 18z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            style="margin-left: 14px;"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-12c4qo2-MuiTypography-root"
            />
          </div>
        </section>
      </div>
      <div
        class="sc-gKPRtg lPbDl"
      >
        <div>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            style="align-self: stretch; width: 100%;"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1vk6yy5-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="content"
              id="content-label"
            >
              Your comment
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-15v7luy-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <textarea
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                id="content"
                required=""
                rows="6"
                style="height: 0px; overflow: hidden;"
              />
              <textarea
                aria-hidden="true"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                tabindex="-1"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Your comment
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-jSUZER dZJquO"
    >
      <button
        aria-label="Post"
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-19lulel-MuiButtonBase-root-MuiButton-root"
        style="min-width: 128px;"
        tabindex="0"
        type="button"
      >
        Post
        <span
          class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ChevronRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
